mne/beamformer/_compute_beamformer.py

```diff
@@ -103,6 +102,20 @@ def _prepare_beamformer_input(info, forward, label=None, pick_ori=None,
             orient_std)
 
 
+def _sym_inv(x, reduce_rank):
+    """Symmetric inversion with optional rank reduction."""
+    s, u = np.linalg.eigh(x)
+    # mimic default np.linalg.pinv behavior
+    cutoff = 1e-15 * s[:, -1][:, np.newaxis]
+    s[s <= cutoff] = np.inf
+    if reduce_rank:
+        # These are ordered smallest to largest, so we set the first one
+        # to inf -- then the 1. / s below will turn this to zero, as needed.
+        s[:, 0] = np.inf
+    s = 1. / s
+    return np.matmul(u * s[:, np.newaxis], u.transpose(0, 2, 1))
+
+
```

Comment:
Okay, I realized that we are actually not implementing *leadfield rank reduction*. We are reducing the rank of the denominator of the beamformer formula (I think way back in 2017 this was introduced as a bug fix with sphere models, IIRC ...). To have a leadfield rank reduction, we would need to supply the leadfield matrix before it gets multiplied with anything else ...

Reply:
Feel free to push commits with comments if you think it would help. All I tried to do was make it so that the same `reduce_rank` behavior that was in `master` was implemented for all other modes (except diag inversion where it does not make sense / is not safe).

Reply:
(i.e., I did not think to hard about actual beamformer math, just generalizing the rank-deficient pseudo-inversion that was in master)

Reply:
Hi @larsoner this was definitely implemented "wrongly" before you started on this ... With wrong not being the right word either: we just *named* it wrong - what we do works, but is not leadfield rank reduction I think. Not sure on how to proceed, we could rename the parameter? ... @wmvanvliet @agramfort ?

Reply:
`reduce_rank` as a param name is at least reasonable because it does reduce the rank during (pseudo)inversion. There is a cost to deprecation if we rename it. So my vote would just be to update the docstring description to be clearer about what it does

Reply:
Updated the docstring, let me know if you think it's enough...

Reply:
I think the docstring looks fine now. 
The thing here is: we know it works (we tested with phantom data some time ago), but it is AFAIK not part of the literature etc. What people usually do is reducing the rank of the lead field - at this point, we do not know whether one methods outperforms the other (I am looking at @wmvanvliet here ... should we dare to add that into the simulation?), so it might be reasonable to also introduce the lead field rank reduction. Should we give the lead field rank reduction a different name then (and change docstring for `reduce_rank` to reflect the denominator approach) or should we move the denominator rank reduction to a new place (invent a new parameter name) and implement the lead field in place of `reduce_rank`?

My pro :+1: for keeping `reduce_rank`: adding `reduce_rank_fwd` or something alike would be very explicit.
My contra :-1:: maybe users already paired `reduce_rank` with lead field rank reduction both in their head and their scripts. (_Note_: lead field rank reduction is something that is often done per default for MEG data and not only sphere head models, e.g. per default in FieldTrip.)

Reply:
I can add it to the simulation. At the very least, we can see if rank reducing the denominator is equivalent to rank reducing the leadfield. The current implementation of `reduce_rank` works very well in the simulations in cases where no weight normalization is applied.

Reply:
>  so it might be reasonable to also introduce the lead field rank reduction. Should we give the lead field rank reduction a different name then (and change docstring for reduce_rank to reflect the denominator approach) or should we move the denominator rank reduction to a new place (invent a new parameter name) and implement the lead field in place of reduce_rank?

Yes I think we should add the lead field method later. We can even treat `reduce_rank=True` as a backward compatible alias for the more explicit `reduce_rank='denominator'` (this PR) and later we add `reduce_rank='leadfield'` mode or whatever we want to call it. I think as long as we're explicit in the docstrings about what is happening it should be fine.

Reply:
Oh, I like this version @larsoner. I will coordinate with Marijn to get a lead field rank reduction into the simulations to see if it behaves differently. We can then open a PR if it does, WDYT?

Reply:
I say let's merge this as is, then work on the lead field rank reduction in a subsequent PR, yeah

Reply:
IMO This PR can be merged if people are waiting on it. In the simulations, it performs well.

Of course, we'll revisit all the parameters after our beamformer simulation is complete. The idea is still to reduce the number of parameters to a manageable number. Hope you'll not be offended if we end up undoing some of your work @larsoner.

Reply:
Yes if you kill this form of rank reduction down the line it's fine by me

Reply:
Awesome, I just agreed with @wmvanvliet that I will start implementing the lead field rank reduction and he will test with simulations once done. I will follow the parameter names we agreed on. 

What do we do with *test* in this one?

Reply:
I'm not quite sure what you're asking. Do you mean what do we do with the tests in this PR? Or are you asking how we add tests for lead field rank reduction once you implement it?

Reply:
Ahh okay, sounds good :)

Reply:
Sorry for the confusion @larsoner :smiley: